src/components/ProtectedRoute.tsx

```diff
@@ -0,0 +1,28 @@
+import React, { useState, useEffect } from 'react';
+import { Navigate, Outlet } from 'react-router-dom';
+import { fetchUserProfile } from '../api/user';
+
+const ProtectedRoute: React.FC = () => {
+  const [isAuthenticated, setIsAuthenticated] = useState<boolean | null>(null);
+  useEffect(() => {
+    const checkAuthentication = async () => {
+      try {
+        const userProfile = await fetchUserProfile();
+        if (userProfile === null) {
+          setIsAuthenticated(false);
+        } else {
+          setIsAuthenticated(true);
+        }
+      } catch {
+        setIsAuthenticated(true);
```

Comment:
Why set isAuthenticated to true on a thrown exception?

Reply:
Since `/user/me` is our only endpoint to verify authentication, I implemented the fetchUserProfile() function to return null if the response is 401 and throw an error for any other response besides 200. This is why I set isAuthenticated to true when an exception is thrown. The idea is to assume the user is still authenticated and avoid logging them out due to temporary issues like network errors. I have an error page on the profile page to handle cases when fetching user data fails.

Reply:
It's not a good idea to just assume anything. Instead you could handle this by either having a sort of mechanism to re-check after some period of time (make an additional /user/me call) and then logging out the user to avoid prolonging the issue if it keeps failing. Cause if the server is continuously down, what's the point of the user being logged in, they can't use the app properly anyways. 
This is usually handled by a refresh token mechanism, which you're gonna implement separately so for now let's keep it this why. Just remember it so you can fix this later on.